vllm/timings/timer.py

```diff
@@ -0,0 +1,75 @@
+import time
+from contextlib import contextmanager
+from threading import RLock
+
+import numpy
+
+__all__ = ["Timer"]
+
+
+class Timer:
+    """Timer to log timings during inference requests. Should be used through 
+        timings.utils.get_singleton_manager()
+    """
+    _instance = None
+
+    def __init__(self, avg_after_iterations: int, log: bool):
+        """
+        :param avg_after_iterations: the number of iterations that have to occur
+            before the time measurements are averaged and displayed to the user
+        :param log: whether or not time logging is enabled
+        """
+        self.log = log
+        self.measurements = {}
+        self.avg_after_iterations = avg_after_iterations
+        self.current_iter = 0
+        self.lock = RLock()
+
+    @contextmanager
+    def time(self, func_name: str):
+        """
+        Time the given function, add the time to the dictionary of lists 
+        tracking each function and calculate the average, if 
+        self.avg_after_iterations number of iterations have passed.
+
+        :param func_name: name of the function that will be used to log the 
+            measurements
+        """
+        start = time.time()
+        try:
+            yield
+        finally:
+            end = time.time()
+            with self.lock:
+                self._update(func_name, end - start)
+                self._compute_average()
+
+    def _update(self, func_name: str, time_elapased: float):
+        """
+        Update the dictionary of measurements. 
+
+        :param func_name: name of the function that will be used to log the 
+            measurements
+        :param time_elapsed: time taken for function execution  
+        """
+        self.current_iter += 1
```

Comment:
should current_iter be tied to a func name?

Reply:
Ya that's what my TODO/remaining question was about.  Had a function to do this initially but wasn't sure if that's what we wanted

Reply:
As well as if we'd like to clear measurements after x iterations or not

Reply:
If server is used, then its inevitable for the dict size to increase. 
If prometheus is in the scope, you can just "print" the values to prometheus and get averages. And yes its compatible with async or sync -- or just wait until async recordings are awaited and print to the sync prometheus.
We can also "print" per func_name or whatever label you give it and it can generate a visualization with respect to that label. 
For regular dict recordings, there is a way to generate a histogram of run_times so that will be easier to visualize also.

Reply:
Updated to use a counter for each function. Not clearing any of the calculated measurements so we can update based on whatever we decide to do with all the logged times.